### PR TITLE
Create jsbinctl and refactor Portfolio to use it

### LINKED ISSUE #521: Create JS Bin build system
This issue is to create a JS Bin build system so we can easily update linked JS Bins in workshops.

Here are my current thoughts on doing this:

- JS Bin has an import from GitHub gist feature (where you just go to `https://jsbin.com/<gist_id>`
- We could have a directory for each JS Bin example with files in it and a script to upload each directory as a Gist and programatically update links to `https://jsbin.com/<gist_id>` in the workshops

### PR DESCRIPTION
**Pull request by zachlatta**
This PR creates `jsbinctl`, a super neat-o utility for managing our JS Bin examples. Its README (included in this pull request) should give better, more complete details on how it works.

The one downside to this pull request: given the nature of how reference links work in Markdown (for those unfamiliar, see the _reference_ portion of https://daringfireball.net/projects/markdown/syntax#link), workshops that use this build system won't be able to have JS Bin links open in new tabs. Given how difficult it currently is to edit workshops without the build system, I think this is a fair tradeoff until we figure out a better solution.

Closes #521.

/cc @MaxWofford @paked

### CODE REVIEW

**Comment by MaxWofford** [comment]
@zachlatta We can preface workshops by saying that CTRL+clicking on a link will open it in a new tab.

**Comment by zachlatta** [comment]
Ah, good call @MaxWofford. I like that idea.

What if we submit pull requests to add that after this is merged?

File: workshops/lib/jsbinctl/README.md [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+```md
+Let's build a plumbus. By the end of this workshop, we're going to end up with a
+plumbus that looks like [this][plumbus_prototype].
+
+[plumbus_prototype]: https://gist.jsbin.com/<gist_id>
+```
+
+You can get the `<gist_id>` to use in the link by running `jsbinctl upload
+examples/<directory_name>`.
```
Comment (zachlatta) [inline workshops/lib/jsbinctl/README.md]: Uh oh, this should be updated. The `upload` command actually returns the full JS Bin link, not just the gist id.

File: workshops/lib/jsbinctl/jsbinctl [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        Yes ) break;;
+        No ) exit;;
+      esac
+    done
+
+    # match groups:
+    #
+    # 1) [example_name]:
+    # 2) https://gist.jsbin.com/old
+    # 3: ?output,html
```
Comment (zachlatta) [inline workshops/lib/jsbinctl/jsbinctl]: Consistency! Use `)`.

File: workshops/lib/jsbinctl/jsbinctl [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+      case $yn in
+        Yes ) break;;
+        No ) exit;;
+      esac
+    done
+
+    # match groups:
+    #
+    # 1) [example_name]:
+    # 2) https://gist.jsbin.com/old
```
Comment (zachlatta) [inline workshops/lib/jsbinctl/jsbinctl]: `old` is not a very good example. Maybe change it to `oldgistid`?

File: workshops/portfolio/README.md [hunk 4]
```diff
@@ -13,12 +13,9 @@ Your final design looks like this:
...
 > ![](img/final_1.png)
 
-Here is a link to a
-<a href="http://output.jsbin.com/fugoki/2" target="_blank">live demo</a>
+Here is a link to a [live demo][final_output] live demo.
 
-And here is the
-<a href="http://jsbin.com/fugoki/2/edit" target="_blank">final code</a>
-for the live demo.
+And here is the [final code][final_output] for the live demo.
```
Comment (zachlatta) [inline workshops/portfolio/README.md]: Does this properly link to the code? Or just the JS Bin output?

**Comment by paked** [comment]
I think there is another downside... People have to actually run commands, which aren't that well documented (eg. can't google `jsbinctl fails no host found`, etc.). Maybe we should look hooking this up to the circle-ci in order to remove that dependency?

**Comment by zachlatta** [comment]
I'm not sure if wiring up CircleCI (or @prophetorpheus) to automatically update links would make sense, since it'd need to submit a pull request to update the links. Every time someone submits a pull request to upload a JS Bin example, both that and the pull requests from CircleCI would need to be merged, which sounds like extra coordination and complexity.

Instead, I propose that we include a "how to use `jsbinctl`" in our "How to contribute to workshops" guide. And if someone submits a pull request that changes a JS Bin example without running `jsbinctl`, we can manually just submit another PR that updates the links.

**Comment by zachlatta** [comment]
Also, just for the sake of being a proponent for this pull request, right now it's basically impossible for an external contributor to edit our JS Bin examples (they'd need to manually copy and paste the code from an existing JS Bin example into a new JS Bin that they have to manually create). This pull request at least makes that process much more accessible.

**Comment by zachlatta** [comment]
Okay, this should be good to go.

I propose we merge this as as (without squashing) because each commit is a logical change that adheres to our commit guidelines.

Now it's up to you @jonleung.

**Comment by jonleung** [comment]
Re:
> right now it's basically impossible for an external contributor to edit our JS Bin examples (they'd need to manually copy and paste the code from an existing JS Bin example into a new JS Bin that they have to manually create

> _Zach Latta_

Actually, if you open someone else's existing JS Bin and just do `File > Save`, it automatically clones and saves the JS Bin.

Re:
> I'm not sure if wiring up CircleCI (or @prophetorpheus) to automatically update links would make sense, since it'd need to submit a pull request to update the links. Every time someone submits a pull request to upload a JS Bin example, both that and the pull requests from CircleCI would need to be merged, which sounds like extra coordination and complexity.

Just to clarify @zachlatta, do you mean that:

1. Person A submits a pull request that modifies the source code of a workshop
2. Circle CI submits a pull request to modify your pull request and leaves a comment in Person A's original pull request
3. Person A reviews Circle CI's pull request in and merges Circle CI's pull request in if person A finds it satisfactory

_Then everything afterwards is normal:_

1. Person A assigns person B to reviews the now combined pull requests
2. Person B reviews the pull request and merges it in as per normal

@paked were you thinking about something like the above ^

**Comment by jonleung** [comment]
Issues: I'll add issues to this comment as I notice them in my review and leave another comment that I've left a comment (so that everyone gets notifications):

- [ ] It is not clear how to install `jsbinctl` from the documentation. I am actually not sure where the best place to put `jsbinctl` is and a beginner definitely will not

**Comment by jonleung** [comment]
Added another checkbox

**Comment by jonleung** [comment]
I'm currently getting this error @zachlatta:

```
$ jsbinctl update plumbus_prototype
readlink: illegal option -- f
usage: readlink [-n] [file ...]
usage: basename string [suffix]
       basename [-a] [-s suffix] string [...]
readlink: illegal option -- f
usage: readlink [-n] [file ...]
I'm about to update all  references to the new JS Bin URL (Error: Is a directory @ io_fread - /Applications) in examples. Let me continue?
1) Yes
2) No
#? 
```

**Comment by jonleung** [comment]
Also, it would be nice to be able to do

```
jsbinctl README.md
```

and it would automatically update all the JS Bin links

**Comment by paked** [comment]
@zachlatta We need a shell one liner to install `jsbinctl`.

**Comment by zachlatta** [comment]
Alright, I've made `jsbinctl` work on OS X and I've added (bad) instructions for
running it. I think we can improve the instructions later.

I don't think we should have a bot automatically submit pull requests for this
because the bot will be complicated to write and adds additional complexity
since it's a black box to the contributor. Despite the (potential) added
difficulty, I think `jsbinctl` isn't very difficult to run and we can improve it
to make the exprience for the contributor seamless.

I'm gonna assign @MaxWofford to review this since @jonleung is in San Diego and
we agreed that @MaxWofford and I will take all of @jonleung's pull requests.

**Pull request merged**